### DISCUSSION TITLE
Stability of the STAC requests

### DISCUSSION THREAD

**Discussion by lukasValentin** [microsoft/PlanetaryComputer#222] [General]
Hi Planetary Computer folks,

first of all many thanks for putting effort into this super nice platform - I like it quiet a lot and use it for most of my research work as a PhD student.

Together with my colleagues at the [Group of Crop Science at ETH Zurich](https://kp.ethz.ch/) and the [EOA-team at Agroscope Reckenholz](https://www.eoa-team.net/) I have been developing the open-source Python package [EOdal](https://github.com/EOA-team/eodal/tree/master), which we also run to fetch Sentinel-2 data from Planetary Computer (see [here](https://github.com/EOA-team/eodal/blob/master/scripts/sentinel2_mapper.py) for an example).

Since we have powerful local hard-ware infrastructure we usually make the queries to STAC from our own machines to fetch the Sentinel-2 data, which is then input into our models required for, e.g., the retrieval of functional crop traits.
While this works fantastically *most* of the time, in some cases we also get an error message when sending our requests to the STAC endpoint:

```html
<!DOCTYPE html PUBLIC '-//W3C//DTD XHTML 1.0 Transitional//EN' 'http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd'><html xmlns='http://www.w3.org/1999/xhtml'><head><meta content='text/html; charset=utf-8' http-equiv='content-type'/><style type='text/css'>body {font-family:Arial; margin-left:40px; }img  { border:0 none; }#content { margin-left: auto; margin-right: auto }#message h2 { font-size: 20px; font-weight: normal; color: #000000; margin: 34px 0px 0px 0px }#message p  { font-size: 13px; color: #000000; margin: 7px 0px 0px0px}#errorref { font-size: 11px; color: #737373; margin-top: 41px }</style><title>Service unavailable</title></head><body><div id='content'><div id='message'><h2>Our services aren't available right now</h2><p>We're working to restore all services as soon as possible. Please check back soon.</p></div><div id='errorref'><span>0IEJvZAAAAAAwuMY3VRAMQomIQ2YvOlz0WlJIRURHRTEzMjIAOTI3YWJmYTYtMTlmNi00YWYxLWEwOWQtYzk1OWQ5YTFlNjQ0</span></div></div></body></html>
```
This happens seemingly at random (i.e., you can send the same requests 10 times and X times it fails while 10-X time it passes). We query small subsets of data (e.g., a single tile for a single month) as we are aware that there must be some limitations on the server side.

However, it would still be nice if those errors would occur less frequently as it arguably impacts the stability of our programs and requires a lot of error handling.

So I was wondering if there is any possibility to decrease the frequency at which these errors occur? For our applications it would be also fine do work asynchronously, i.e., it does not matter if we have to wait some seconds or not until our request is processed.

I am looking forward to hearing back from you.

Cheers,
Lukas

**Comment by TomAugspurger** [comment]
Thanks. A recent update has introduced more of these errors that we're still working through (https://github.com/microsoft/PlanetaryComputer/issues/217). In the meantime, you might consider adding retries to your requests to the STAC API. https://github.com/stac-utils/pystac-client/issues/520 has some discussion on that, if you're using pystac-client.

**Reply by lukasValentin** [reply]
Thanks for the hint - indeed we are using pystac-client for our requests

**Comment by jessjaco** [comment]
I've ended up sprinkling [`@retry()` decorators](https://pypi.org/project/retry/) over everything that reads or writes network data.
